src/isl.rs

```diff
@@ -0,0 +1,299 @@
+                // inserts ISLCoreType as a Type into type_store
+                Ok(context.borrow_mut().add_named_type(
+                    &format!("{:?}", ion_type),
+                    TypeDefinition::new(Some(format!("{:?}", ion_type)), vec![]),
+                    type_store,
+                ))
+            }
+            IslTypeRef::NamedType(alias) => {
+                let borrow_context = context.borrow_mut();
+                // verify if the AliasType actually exists in the type_store or throw an error
+                match borrow_context.get_type_id_by_name(alias, type_store) {
+                    Some(type_id) => Ok(type_id.to_owned()),
+                    None => match borrow_context.get_parent() {
+                        Some(parent) => {
+                            // if it is a self referencing type resolve it using parent information from type_store
+                            if parent.0.eq(alias) {
+                                Ok(parent.1)
+                            } else {
+                                unresolvable_schema_error(format!(
+                                    "Could not resolve type reference: {:?} does not exist",
+                                    alias
+                                ))
+                            }
+                        }
+                        None => unresolvable_schema_error(format!(
+                            "Could not resolve type reference: {:?} does not exist",
+                            alias
+                        )),
+                    },
+                }
+            }
+            IslTypeRef::AnonymousType(isl_type) => {
+                let type_id = context.borrow().get_total_types();
+                let type_def = TypeDefinition::parse_from_isl_type_and_update_type_store(
+                    isl_type, type_store, context,
+                )?;
+                // get the last added anonymous type's type_id for given anonymous type
+                Ok(type_id)
+            } //TODO: add a check for ImportType type reference here
+        }
+    }
+}
+
+#[cfg(test)]
+mod isl_tests {
+    use super::*;
+    use ion_rs::value::owned::OwnedElement;
+    use ion_rs::value::reader::element_reader;
+    use ion_rs::value::reader::ElementReader;
+    use rstest::*;
+
+    // helper function to be used by isl tests
+    fn load(text: &str) -> OwnedElement {
+        element_reader()
+            .read_one(text.as_bytes())
+            .expect("parsing failed unexpectedly")
+    }
+
+    #[rstest(
+    isl_type1,isl_type2,
+    case::type_constraint_with_anonymous_type(
+        (&load(r#" // For a schema with single anonymous type
+            {type: int}
+        "#)).try_into().unwrap(),
+        IslType::new(None, vec![IslConstraint::Type(IslTypeRef::CoreIslType(IonType::Integer))])
+    ),
+    case::type_constraint_with_named_type(
+        (&load(r#" // For a schema with named type 
+            { name: my_int, type: int }
```

Comment:
Minor nitpick: according to the spec, a named type is required to be annotated with `type::`. (Without that annotation, it's an inline type, and the `name` field is just "open content".) I don't know if it makes any functional difference in this test, but it's a good idea to make sure that the test inputs conform to the ISL spec (unless we're testing an unhappy case, of course).

Reply:
Yes, there needs to be type:: annotation check for named type as this has been already checked in `schema_from_elements` method defined in `system` I haven't added a check here. Also, currently both anonymous/named types are defined by same struct `TypeDefinition`. But in a separate PR with solution #25 I will fix this by adding a check for named types specifically.